algosdk/future/transaction.py

```diff
@@ -2906,6 +2909,103 @@ def __eq__(self, other):
         return False
 
 
+class StateProofTxn(Transaction):
+    """
+    Represents a state proof transaction
+
+    Arguments:
+        sender (str): address of the sender
+        state_proof (dict())
+        state_proof_message (dict(), optional)
+        state_proof_type (str, optional): state proof type
+        sp (SuggestedParams): suggested params from algod
+        receiver (str): address of the receiver
+        amt (int): amount in microAlgos to be sent
+        close_remainder_to (str, optional): if nonempty, account will be closed
+            and remaining algos will be sent to this address
+        note (bytes, optional): arbitrary optional bytes
+        lease (byte[32], optional): specifies a lease, and no other transaction
+            with the same sender and lease can be confirmed in this
+            transaction's valid rounds
+        rekey_to (str, optional): additionally rekey the sender to this address
+
+
+    Attributes:
+        sender (str)
+        sprf (dict())
+        sprfmsg (dict())
+        sprf_type (str)
+        fee (int)
+        first_valid_round (int)
+        last_valid_round (int)
+        note (bytes)
+        genesis_id (str)
+        genesis_hash (str)
+        group (bytes)
+        receiver (str)
+        amt (int)
+        close_remainder_to (str)
+        type (str)
+        lease (byte[32])
+        rekey_to (str)
+    """
+
+    def __init__(
+        self,
+        sender,
+        sp,
+        state_proof,
+        state_proof_message=None,
+        state_proof_type=None,
+        note=None,
+        lease=None,
+        rekey_to=None,
+    ):
+        Transaction.__init__(
+            self, sender, sp, note, lease, constants.stateproof_txn, rekey_to
+        )
+
+        self.sprf_type = state_proof_type
+        self.sprf = state_proof
+        self.sprfmsg = state_proof_message
+
+    def dictify(self):
+        d = dict()
+        if self.sprf_type:
+            d["sptype"] = self.sprf_type
+        if self.sprfmsg:
+            d["spmsg"] = self.sprfmsg
+        d["sp"] = self.sprf
```

Comment:
Is there a reason why we don't check if self.sprf

Reply:
I was thinking that since it's a stateproof txn type, state_roof can't be None.

Reply:
This is how the stateproof txn is defined on go-algorand
https://github.com/algorand/go-algorand/blob/7de989aa62a01ce9971b48cb1c2ca6cc2f18f3bf/data/transactions/stateproof.go#L31

Looks like the type, the message, and the proof can't be None (not pointers). I think they should be referenced in the same way (i.e should remove the if for all of them I guess)

Reply:
the struct has `omitempty,omitemptyarray`, I thought some of these fields could be empty.